File: payola/web/server/public/javascripts/runtime/s2js/runtime/scala/collection/Map.js
goog.provide('scala.collection.Map');
goog.require('scala.NoSuchElementException');
goog.require('scala.None');
goog.require('scala.Option');
goog.require('scala.Tuple2');
goog.require('scala.UnsupportedOperationException');
goog.require('scala.collection.Iterable');
goog.require('scala.collection.mutable.ListBuffer');
scala.collection.Map = function() {
var self = this;
self.internalJsObject = {};
};
goog.inherits(scala.collection.Map, scala.collection.Iterable);
scala.collection.Map.prototype.foreach = function(f) {
var self = this;

        for (var key in self.internalJsObject) {
            if (self.internalJsObject.hasOwnProperty(key)) {
                f(new scala.Tuple2(key, self.internalJsObject[key]))
            }
        }
    };
scala.collection.Map.prototype.$plus$eq = function(x) {
var self = this;
self.internalJsObject[x._1] = x._2;};
scala.collection.Map.prototype.prepend = function(x) {
var self = this;
(function() {
throw new scala.UnsupportedOperationException('Can\'t prepend to a Map.');
})();
};
scala.collection.Map.prototype.reversed = function() {
var self = this;
return scala.collection.mutable.ListBuffer.empty().$plus$plus(self).reversed();
};
scala.collection.Map.prototype.$minus$eq = function(x) {
var self = this;
delete self.internalJsObject[x];};
scala.collection.Map.prototype.get = function(key) {
var self = this;

        if (s2js.isUndefined(self.internalJsObject[key])) {
            return scala.None;
        } else {
            return new scala.Some(self.internalJsObject[key]);
        }
    };
scala.collection.Map.prototype.getOrElse = function(key, $default) {
var self = this;
return (function($selector$1) {
if (s2js.isInstanceOf($selector$1, 'scala.Some') && (true)) {
var v = $selector$1.productElement(0);
return v;
}
if ($selector$1 === scala.None) {
return $default;
}
})(self.get(key));
};
scala.collection.Map.prototype.$apply = function(key) {
var self = this;
return (function($selector$2) {
if ($selector$2 === scala.None) {
return self.$default(key);
}
if (s2js.isInstanceOf($selector$2, 'scala.Some') && (true)) {
var value = $selector$2.productElement(0);
return value;
}
})(self.get(key));
};
scala.collection.Map.prototype.contains = function(key) {
var self = this;
return self.get(key).isDefined();
};
scala.collection.Map.prototype.isDefinedAt = function(key) {
var self = this;
return self.contains(key);
};
scala.collection.Map.prototype.$default = function(key) {
var self = this;
return (function() {
throw new scala.NoSuchElementException(('key not found: ' + key));
})();
};
scala.collection.Map.prototype.put = function(key, value) {
var self = this;
var r = self.get(key);
self.update(key, value);
return r;

};
scala.collection.Map.prototype.update = function(key, value) {
var self = this;
self.$plus$eq(new scala.Tuple2(key, value));
};
scala.collection.Map.prototype.remove = function(key) {
var self = this;
var r = self.get(key);
self.$minus$eq(key);
return r;

};
scala.collection.Map.prototype.getOrElseUpdate = function(key, op) {
var self = this;
return (function($selector$3) {
if (s2js.isInstanceOf($selector$3, 'scala.Some') && (true)) {
var v = $selector$3.productElement(0);
return v;
}
if ($selector$3 === scala.None) {
var d = op;
self.update(key, d);
return d;

}
})(self.get(key));
};
scala.collection.Map.prototype.__class__ = new s2js.Class('scala.collection.Map', [scala.collection.Iterable]);

File: payola/s2js/runtime/client/src/main/scala/s2js/runtime/client/scala/collection/Map.scala
package s2js.runtime.client.scala.collection

import mutable.ListBuffer
import s2js.compiler.{dependency, javascript}

@dependency("scala.Tuple2")
@dependency("scala.Option")
trait Map[A, B] extends Iterable
{
    @javascript("{}")
    var internalJsObject = null

    @javascript("""
        for (var key in self.internalJsObject) {
            if (self.internalJsObject.hasOwnProperty(key)) {
                f(new scala.Tuple2(key, self.internalJsObject[key]))
            }
        }
    """)
    def foreach[U](f: Double => U) {}

    @javascript("self.internalJsObject[x._1] = x._2;")
    def +=(x: Any) {}

    def prepend(x: Any) {
        throw new s2js.runtime.client.scala.UnsupportedOperationException("Can't prepend to a Map.")
    }

    // From TraversableLike
    def reversed: Iterable = {
        (ListBuffer.empty ++ this).reversed
    }

    @javascript("delete self.internalJsObject[x];")
    def -=(x: Any) {}

    @javascript("""
        if (s2js.runtime.client.core.get().isUndefined(self.internalJsObject[key])) {
            return scala.None.get();
        } else {
            return new scala.Some(self.internalJsObject[key]);
        }
    """)
    def get(key: A): Option[B] = None

    // From MapLike
    def getOrElse[B1 >: B](key: A, default: => B1): B1 = get(key) match {
        case Some(v) => v
        case None => default
    }

    // From MapLike
    def apply(key: A): B = get(key) match {
        case None => default(key)
        case Some(value) => value
    }

    // From MapLike
    def contains(key: A): Boolean = get(key).isDefined

    // From MapLike
    def isDefinedAt(key: A) = contains(key)

    // From MapLike
    def default(key: A): B = throw new s2js.runtime.client.scala.NoSuchElementException("key not found: " + key)

    // From mutable.MapLike
    def put(key: A, value: B): Option[B] = {
        val r = get(key)
        update(key, value)
        r
    }

    // From mutable.MapLike
    def update(key: A, value: B) {
        this += ((key, value))
    }

    // From mutable.MapLike
    def remove(key: A): Option[B] = {
        val r = get(key)
        this -= key
        r
    }

    // From mutable.MapLike
    def getOrElseUpdate(key: A, op: => B): B = {
        get(key) match {
            case Some(v) => v
            case None => val d = op; this(key) = d; d
        }
    }

    @javascript("""
        var result = scala.collection.mutable.ListBuffer.get().$apply();
        for (var key in self.internalJsObject) {
            if (key === 'length' || !widthRange.hasOwnProperty(key)) {
                continue;
            }
            result.$plus$plus$eq(widthRange[key]);
        }
        return result;
    """)
    def values(): Iterable = null

    @javascript("""
        var result = scala.collection.mutable.ListBuffer.get().$apply();
        for (var key in self.internalJsObject) {
            if (key === 'length' || !self.internalJsObject.hasOwnProperty(key)) {
                continue;
            }
            result.$plus$eq(key);
        }
        return result;
    """)
    def keys(): Iterable = null
}
